src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/RetryLogic/RetryLogicCounterTest.cs

```diff
@@ -0,0 +1,192 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Microsoft.Data.SqlClient.ManualTesting.Tests
+{
+    public class RetryLogicCounterTest
+    {
+        [ConditionalTheory(typeof(DataTestUtility), nameof(DataTestUtility.AreConnStringsSetup))]
+        [InlineData("ExecuteScalarAsync", 3)]
+        [InlineData("ExecuteReaderAsync", 3)]
+        [InlineData("ExecuteXmlReaderAsync", 3)]
+        [InlineData("ExecuteNonQueryAsync", 3)]
+        public async void ValidateRetryCount_SqlCommand_Async(string methodName, int numOfTries)
+        {
+            RetryLogicTestHelper.SetRetrySwitch(true);
+
+            ErrorInfoRetryLogicProvider _errorInfoRetryProvider = new(
+                SqlConfigurableRetryFactory.CreateFixedRetryProvider(new SqlRetryLogicOption()
+                { NumberOfTries = numOfTries, TransientErrors = new[] { 50000 } }));
+
+            using var connection = new SqlConnection(DataTestUtility.TCPConnectionString);
+            connection.Open();
+
+            using SqlCommand cmd = connection.CreateCommand();
+            cmd.RetryLogicProvider = _errorInfoRetryProvider;
+            cmd.CommandText = "THROW 50000,'Error',0";
+
+            _errorInfoRetryProvider.CallCounter = 0;
+            try
+            {
+                switch (methodName)
+                {
+                    case "ExecuteScalarAsync":
+                        await cmd.ExecuteScalarAsync();
+                        break;
+                    case "ExecuteReaderAsync":
+                        {
+                            using var _ = await cmd.ExecuteReaderAsync();
+                            break;
+                        }
+                    case "ExecuteXmlReaderAsync":
+                        {
+                            using var _ = await cmd.ExecuteXmlReaderAsync();
+                            break;
+                        }
+                    case "ExecuteNonQueryAsync":
+                        await cmd.ExecuteNonQueryAsync();
+                        break;
+                    default:
+                        break;
+                }
+                Assert.False(true, "Exception did not occur.");
+            }
+            catch
+            {
+                Assert.Equal(numOfTries, _errorInfoRetryProvider.CallCounter);
+            }
+            RetryLogicTestHelper.SetRetrySwitch(false);
```

Comment:
Could you move the `RetryLogicTestHelper.SetRetrySwitch(true);` into the try-block and wrap this line with a finally-block to be sure it will be called?